heron/tools/apiserver/src/java/com/twitter/heron/apiserver/resources/FileResource.java

```diff
@@ -0,0 +1,153 @@
+  private static final Logger LOG = LoggerFactory.getLogger(FileResource.class);
+
+  private static final String FILE_SYSTEM_DIRECTORY
+      = "heron.apiserver.http.file.system.directory";
+
+  private static InetAddress ip;
+  private static String hostname;
+
+  static {
+    try {
+      ip = InetAddress.getLocalHost();
+      hostname = ip.getHostName();
+    } catch (UnknownHostException e) {
+      LOG.info("Failed or resolve IP address of localhost");
+    }
+  }
+
+  @POST
+  @Path("/upload")
+  @Consumes(MediaType.MULTIPART_FORM_DATA)
+  public Response uploadFile(
+      @FormDataParam("file") InputStream uploadedInputStream,
+      @FormDataParam("file") FormDataContentDisposition fileDetail) {
+
+    Config config = createConfig();
+
+    if (uploadedInputStream == null) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+
+    }
+    if (fileDetail == null) {
+      return Response.status(Response.Status.BAD_REQUEST).build();
+    }
+
+    String uploadDir = config.getStringValue(FILE_SYSTEM_DIRECTORY);
+
+    String fileName = UUID.randomUUID() + "-" + fileDetail.getFileName();
+
+    String uploadedFileLocation
+        = uploadDir + "/" + fileName;
+
+    // save it
+    writeToFile(uploadedInputStream, uploadedFileLocation);
+
+    String uri = String.format("http://%s:%s/api/v1/file/download/%s",
+        (hostname != null) ? hostname : ip, getPort(), fileName);
+
+    return Response.status(Response.Status.OK).entity(uri).build();
+  }
+
+  @GET
+  @Path("/download/{file}")
+  public Response downloadPdfFile(final @PathParam("file") String file) {
+    Config config = createConfig();
+    String uploadDir = config.getStringValue(FILE_SYSTEM_DIRECTORY);
+    String filePath = uploadDir + "/" + file;
+    if (!new File(filePath).exists()) {
+      LOG.debug("Download request file " + file + " doesn't exist at " + uploadDir);
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+    StreamingOutput fileStream = new StreamingOutput() {
+      @Override
+      public void write(java.io.OutputStream output) throws IOException {
+          java.nio.file.Path path = Paths.get(filePath);
+          byte[] data = Files.readAllBytes(path);
+          output.write(data);
+          output.flush();
+      }
+    };
+    return Response
+        .ok(fileStream, MediaType.APPLICATION_OCTET_STREAM)
+        .header("content-disposition", "attachment; filename = " + file)
+        .build();
+  }
+
+  private Config createConfig() {
+    final Config.Builder builder = Config.newBuilder().putAll(getBaseConfiguration());
+    builder.put(Key.VERBOSE, Logging.isVerbose());
+    return Config.toLocalMode(builder.build());
+  }
+
+  // save uploaded file to new location
+  private void writeToFile(InputStream uploadedInputStream,
+                           String uploadedFileLocation) {
+    File file = new File(uploadedFileLocation);
+    file.getParentFile().mkdirs();
+
+    try {
```

Comment:
Maybe use a util here? FileHelper.copy(in, path)?

Reply:
I can move the function to FileHelper but the existing copy only copies files already on disk